Taweret/mix/linear.py

```diff
@@ -68,50 +65,46 @@ def __init__(self, model_1, model_2, x_exp, y_exp, y_err, method='sigmoid',n_mod
-        #print(self.L1)
-        #print(self.L2)
-
-    def mix_loglikelihood(self, mixture_params : np.ndarray, model_1_param=np.array([]), model_2_param=np.array([])) -> float:
+    def mix_loglikelihood(self, mixture_params : np.ndarray, model_params=[]) -> float:
         """
         log likelihood of the mixed model given the mixing function parameters
 
         Parameters
         ----------
         mixture_params : np.1darray
             parameter values that fix the shape of mixing function
-        model_1_param: np.1darray
-            parameter values in the model 1
-        model_2_param: np.1darray
-            parameter values  in the model 2
+        model_params: list[np.1darray]
+            list of model parameters for each model, note that different models can take different
+            number of params
         """
 
-        W = mixture_function(self.method, self.x_exp, mixture_params)
-        W_1 = np.log(W + eps)
-        W_2 = np.log(1 - W + eps)
-        if self.n_model_1==0 and self.n_model_2==0:
-            complete_array=np.append(W_1+self.L1, W_2+self.L2)
+        # check that list of models and mixture_params has same length
+        if len(self.models) != len(mixture_params) and len(self.nargs_for_each_model) != 0:
+            raise Exception('linear_mix.mix_loglikelihood: mixture_params has wrong length')
+
+        # here, the mixture_function should support dirchlet distribution which is the most
+        # general distribution for the symplex defined be mixture_params
+
+        # return weights for different models and take logs
+        weights = mixture_function(self.method, self.x_exp, mixture_params)
+        log_weights = np.array([np.log(weight + eps) for i, weight in enumerate(weights)])
+
+        # calculate log likelihoods
+        if len(self.nargs_for_each_model) == 0:
+            log_likelis = np.array(
+                [log_likelihood_elementwise(model, self.x_exp, self.y_exp, self.y_err) + log_weight
+                 for model, log_weight in zip(self.models, log_weights)])
         else:
-            L1 = log_likelihood_elementwise(self.model_1, self.x_exp, self.y_exp, self.y_err, model_1_param)
-            L2 = log_likelihood_elementwise(self.model_2, self.x_exp, self.y_exp, self.y_err, model_2_param)
-            complete_array=np.append(W_1+L1, W_2+L2)
-        #print(complete_array)
-        total_sum = 0
-        for i in range(0,len(complete_array)-1):
-            if i==0:
-                total_sum=np.logaddexp(complete_array[i],complete_array[i+1])
-            else:
-                total_sum=np.logaddexp(total_sum, complete_array[i+1])
-            #print(total_sum)
+            log_likelis = np.array(
+                [log_likelihood_elementwise(model, self.x_exp, self.y_exp, self.y_err, params) + log_weight
+                 for model, params, log_weight in zip(self.models, model_params, log_weights)])
+
+        total_sum = np.log(np.sum(np.exp(log_likelis)))
```

Comment:
Taking the exponential and then again taking the log will amplify the numerical error. Can this be done with an inbuilt NumPy method like `logaddexp`

Reply:
Numpy seems to have a method `np.logaddexp.reduce(arr)` that applies the method to a list

Reply:
Oh wow. That's really great. I couldn't find documentation on this. Can you post it here?

Reply:
The API originates from here: https://numpy.org/doc/stable/reference/generated/numpy.ufunc.reduce.html 
Since `np.logaddexp` is a `np.ufunc` it automatically has this function available to it

Reply:
`log_likelihood_elementwise` calculates likelihood for each value of input parameter 'x'. In global mixing you don't care about the `log_likelihood_elementwise` but you mix the total likelihood for each model with weights. You will need to change this. Can it be modified in a general way so it will work for both global and local mixing methods?

Reply:
Also, the modification here does not work for local mixing methods. Because now you have modified the format of the output of weights (`mixing_function`). Maybe it's best if we decouple local and global mixing and introduce multi_modal_linear mixing as an entirely new mixing method (a seperate file to linear.py).